contentpal/src/main/java/org/dmfs/android/contentpal/operations/DelegatingInsertOperation.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2017 dmfs GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.dmfs.android.contentpal.operations;
+
+import android.content.ContentProviderOperation;
+import android.support.annotation.NonNull;
+
+import org.dmfs.android.contentpal.InsertOperation;
+import org.dmfs.android.contentpal.Operation;
+import org.dmfs.android.contentpal.SoftRowReference;
+import org.dmfs.android.contentpal.TransactionContext;
+import org.dmfs.optional.Optional;
+
+
+/**
+ * Abstract {@link InsertOperation} which delegates all calls to another operation.
+ *
+ * @author Gabor Keszthelyi
+ */
+public abstract class DelegatingInsertOperation<T> implements InsertOperation<T>
+{
+    private final Operation<T> mDelegate;
+
+
+    public DelegatingInsertOperation(@NonNull Operation<T> delegate)
```

Comment:
Shouldn't this delegate to another `InsertOperation`?

Reply:
We already have cases with `Related` when we need to turn an `Operation` into an `InsertOperation`. I even thought about having a general adapter but that's probably too easy to abuse. So I think what we can do is to create a delegating class in each case we need to compose an `InsertOperation` from `Operation`, to make sure we pay attention that it will be an insert, make sure it's in line with that.
So that's one reason for this class to help make those classes easily.

Does it make sense?

Reply:
So an alternative would be that mentioned general adapter `OperationInsertOperation`.

Does it still have benefits to keep the dedicated `InsertOperation` interface in general?

Reply:
The dedicated interface was added to ensure you can't use update operations when an insert operation is expected. In Android's content providers there are plenty of columns which can (or should) only be set with the insert operation, like the calendar an event belongs to or the account a raw contact belongs to.

An adapter to adapt an `Operation` to an `InsertOperation` would absolutely defeat the purpose (no matter how it's called). So the question is "is the design appropriate?". Can or even should we ensure this sort of operation type safety?

Reply:
I don't know how much `InsertOperation` helps but looking into the 2 examples you mentioned - `CalendarEvent` and `RawContactData`, they do make it clear with the type that they are insert opertations, so I would know that I can't use it for update. So it may make sense to keep them.
But both of them delegate to `Operation` now because of the changed `Related`. So I think this `DelegatingInsertOperation` created in this PR could help with more concisely define them and other `InsertOperation`s.

Reply:
But isn't that a point for not allowing this decorator to accept any `Operation`? Accepting any `Operation` and not just `InsertOperation`s essentially makes it an adapter rather than a decorator, but an adapter that should not exist.

It's like trying to make a truck fit into an underground garage for cars by just mounting a bigger garage door in front of the regular one. Although technically both, trucks and cars, are motor vehicles, that doesn't mean you can add an adapter like that to fit a truck into a car environment (even though for some smaller trucks it might actually work this way).

Reply:
@dmfs What do you say about this?
This would come handy for now `Subtasking` `InsertOperation` in OpenTaskPal as well.

Reply:
Only if we delegate to another `InsertOperation`.

Reply:
Ok, I've modified it to delegate to `InsertOperation`. Added unit test and rebased, ready for review.